File: src/flutter/views/Group_flutter.cpp
/*
  This file is part of KDDockWidgets.

  SPDX-FileCopyrightText: 2019-2023 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
  Author: Sérgio Martins <sergio.martins@kdab.com>

  SPDX-License-Identifier: GPL-2.0-only OR GPL-3.0-only

  Contact KDAB at <info@kdab.com> for commercial licensing options.
*/

/**
 * @file
 * @brief The GUI counterpart of Frame.
 *
 * @author Sérgio Martins \<sergio.martins@kdab.com\>
 */

#include "Group_flutter.h"

#include "kddockwidgets/core/Group.h"
#include "kddockwidgets/core/Stack.h"
#include "kddockwidgets/core/TitleBar.h"
#include "kddockwidgets/core/DockWidget.h"
#include "kddockwidgets/core/DockWidget_p.h"

#include "flutter/ViewFactory_flutter.h"
#include "flutter/Platform_flutter.h"
#include "flutter/views/DockWidget_flutter.h"

#include "Stack_flutter.h"

#include "Config.h"
#include "core/ViewFactory.h"
#include "core/WidgetResizeHandler_p.h"

#include <QDebug>

using namespace KDDockWidgets;
using namespace KDDockWidgets::flutter;

Group_flutter::Group_flutter(Core::Group *controller, View *parent)
    : View_flutter(controller, Core::ViewType::Frame, parent)
    , GroupViewInterface(controller)
{
}

Group_flutter::~Group_flutter()
{
}

void Group_flutter::init()
{
}

int Group_flutter::currentIndex() const
{
    return 0;
}

QSize Group_flutter::minSize() const
{
    const QSize contentsSize = m_group->dockWidgetsMinSize();
    return contentsSize + QSize(0, nonContentsHeight());
}

QSize Group_flutter::maxSizeHint() const
{
    return View_flutter::maxSizeHint();
}

int Group_flutter::nonContentsHeight() const
{
    return 0;
}

QRect Group_flutter::dragRect() const
{
    qFatal("Not implemented");
    return {};
}

File: src/flutter/dart/lib/Group_flutter.dart
/*
  This file is part of KDDockWidgets.

  SPDX-FileCopyrightText: 2019-2023 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
  Author: Sérgio Martins <sergio.martins@kdab.com>

  SPDX-License-Identifier: GPL-2.0-only OR GPL-3.0-only

  Contact KDAB at <info@kdab.com> for commercial licensing options.
*/

import 'package:KDDockWidgets/View_mixin.dart';
import 'package:KDDockWidgets/PositionedWidget.dart';
import 'package:KDDockWidgetsBindings/Bindings.dart' as KDDockWidgetBindings;
import 'package:flutter/material.dart';

import 'DockWidget_flutter.dart';
import 'TabBar_flutter.dart';
import 'TitleBar_flutter.dart';

class Group_flutter extends KDDockWidgetBindings.Group_flutter with View_mixin {
  late final KDDockWidgetBindings.Group m_controller;

  Group_flutter(
      KDDockWidgetBindings.Group? group, KDDockWidgetBindings.View? parent)
      : super(group, parent: parent) {
    m_controller = group!;
    initMixin(this, color: Colors.greenAccent, debugName: "Group");

    print("Group_flutter CTOR");
  }

  Widget createFlutterWidget() {
    return GroupWidget(kddwView, this, key: widgetKey);
  }

  TitleBar_flutter titleBarView() {
    return KDDockWidgetBindings.View_flutter.fromCache(
        m_controller.titleBar().view().thisCpp) as TitleBar_flutter;
  }

  TabBar_flutter tabBarView() {
    return KDDockWidgetBindings.View_flutter.fromCache(
        m_controller.stack().tabBar().view().thisCpp) as TabBar_flutter;
  }

  DockWidget_flutter? dockWidgetView() {
    final dw = m_controller.currentDockWidget();
    if (dw.thisCpp.address != 0) // Add "isNullptr"
      return KDDockWidgetBindings.View_flutter.fromCache(dw.view().thisCpp)
          as DockWidget_flutter;

    print("Group_flutter: No dock widget in the Group!");
    return null;
  }
}

class GroupWidget extends PositionedWidget {
  final Group_flutter groupView;
  GroupWidget(var kddwView, this.groupView, {Key? key})
      : super(kddwView, key: key);

  @override
  State<PositionedWidget> createState() {
    return GroupPositionedWidgetState(kddwView, groupView);
  }
}

class GroupPositionedWidgetState extends PositionedWidgetState {
  final Group_flutter groupView;

  GroupPositionedWidgetState(var kddwView, this.groupView) : super(kddwView);

  @override
  Widget buildContents() {
    final titleBarWidget = groupView.titleBarView().flutterWidget;
    final tabBarWidget = groupView.tabBarView().flutterWidget;
    final dockWidgetWidget = groupView.dockWidgetView()?.flutterWidget;

    return Column(
      children: [
        titleBarWidget,
        tabBarWidget,
        if (dockWidgetWidget != null)
          dockWidgetWidget
        else
          Expanded(child: Container(color: kddwView.m_color))
      ],
    );
  }
}
